authorization/msi-auth-token-provider-jar/src/main/java/com/microsoft/azure/msiAuthTokenProvider/MSICredentials.java

```diff
@@ -0,0 +1,356 @@
+                payload.append("=");
+                payload.append(URLEncoder.encode(this.configForVM.identityId(), "UTF-8"));
+            }
+        } catch (IOException exception) {
+            throw new RuntimeException(exception);
+        }
+
+        int retry = 1;
+        while (retry <= maxRetry) {
+            URL url = new URL(String.format("http://169.254.169.254/metadata/identity/oauth2/token?%s", payload.toString()));
+            //
+            HttpURLConnection connection = null;
+            //
+            try {
+                connection = (HttpURLConnection) url.openConnection();
+                connection.setRequestMethod("GET");
+                connection.setRequestProperty("Metadata", "true");
+                connection.connect();
+                InputStream stream = connection.getInputStream();
+                BufferedReader reader = new BufferedReader(new InputStreamReader(stream, "UTF-8"), 100);
+                String result = reader.readLine();
+                return getMsiTokenFromResult(result, HostType.VIRTUAL_MACHINE);
+            } catch (Exception exception) {
+                int responseCode = connection.getResponseCode();
+                if (responseCode == 410 || responseCode == 429 || responseCode == 404 || (responseCode >= 500 && responseCode <= 599)) {
+                    int retryTimeoutInMs = retrySlots.get(new Random().nextInt(retry));
+                    // Error code 410 indicates IMDS upgrade is in progress, which can take up to 70s
+                    //
+                    retryTimeoutInMs = (responseCode == 410 && retryTimeoutInMs < imdsUpgradeTimeInMs) ? imdsUpgradeTimeInMs : retryTimeoutInMs;
+                    retry++;
+                    if (retry > maxRetry) {
+                        break;
+                    } else {
+                        sleep(retryTimeoutInMs);
+                    }
+                } else {
+                    throw new RuntimeException("Couldn't acquire access token from IMDS, verify your objectId, clientId or msiResourceId", exception);
+                }
+            } finally {
+                if (connection != null) {
+                    connection.disconnect();
+                }
+            }
+        }
+        //
+        if (retry > maxRetry) {
+            throw new RuntimeException(String.format("MSI: Failed to acquire tokens after retrying %s times", maxRetry));
+        }
+        return null;
+    }
+
+    private static MSIToken getMsiTokenFromResult(String result, HostType hostType) throws ParseException {
+        return new MSIToken(getTokenFromResult(result), getTokenTypeFromResult(result), getExpiryTimeFromResult(result, hostType));
+    }
+
+    private static String getTokenFromResult(String result) {
+        int startIndex_AT = result.indexOf(ActiveDirectoryAuthentication.ACCESS_TOKEN_IDENTIFIER)
+                + ActiveDirectoryAuthentication.ACCESS_TOKEN_IDENTIFIER.length();
+
```

Comment:
The only time this is gonna happen is if the result does not hold a token in it. That would happen in a  error scenario... which would fall into the exception category.

But I do like defensive programming...will add the check mate :)

Reply:
thanks :) yes if it happens then it's an error scenario. When MSAL take this over, we can ask there preference on handling this, this may never occur or if happens they might want to throw client side exception (like TokenRetrievalException).